python/ray/tune/integration/mlflow.py

```diff
@@ -0,0 +1,337 @@
+        save_artifact (bool): If set to True, automatically save the entire
+            contents of the Tune local_dir as an artifact to the
+            corresponding run in Mlflow.
+
+    Example:
+
+    .. code-block:: python
+
+        from ray.tune.integration.mlflow import MLFlowLoggerCallback
+        tune.run(
+            train_fn,
+            config={
+                # define search space here
+                "parameter_1": tune.choice([1, 2, 3]),
+                "parameter_2": tune.choice([4, 5, 6]),
+            },
+            callbacks=[MLFlowLoggerCallback(
+                experiment_name="experiment1",
+                save_artifact=True)])
+
+    """
+
+    def __init__(self,
+                 tracking_uri: Optional[str] = None,
+                 registry_uri: Optional[str] = None,
+                 experiment_name: Optional[str] = None,
+                 save_artifact: bool = False):
+        if mlflow is None:
+            raise RuntimeError("mlflow has not been installed. Please `pip "
+                               "install mlflow` to use the MLFlowLogger.")
+
+        self.client = MlflowClient(
+            tracking_uri=tracking_uri, registry_uri=registry_uri)
+
+        # Setup MLFlow experiment.
+        experiment_id = None
+        if experiment_name is None:
+            # If no name is passed in, then check env vars.
+            # First check if experiment_name env var is set.
+            experiment_name = os.environ.get("MLFLOW_EXPERIMENT_NAME")
+            if experiment_name is None:
+                # Then check if experiment_id is set and exists.
+                experiment_id = int(os.environ.get("MLFLOW_EXPERIMENT_ID"))
+                if experiment_id is None or not self.client.get_experiment(
+                        experiment_id):
+                    experiment_id = None
+
+        if experiment_id is None:
+            experiment = self.client.get_experiment_by_name(experiment_name)
+            if experiment is not None:
+                experiment_id = experiment.experiment_id
+            elif experiment_name is not None:
+                experiment_id = self.client.create_experiment(
+                    name=experiment_name)
+            else:
+                raise ValueError("No experiment_name passed, "
+                                 "MLFLOW_EXPERIMENT_NAME env var is not "
+                                 "set, and MLFLOW_EXPERIMENT_ID either "
+                                 "is not set or does not exist. Please "
+                                 "set one of these to use the "
+                                 "MlFlowLoggerCallback.")
+
+
+        self.experiment_id = experiment_id
+        self.save_artifact = save_artifact
+
+        self._trial_runs = {}
+
+    def log_trial_start(self, trial: "Trial"):
+        # Create run if not already exists.
+        if trial not in self._trial_runs:
+            run = self.client.create_run(
```

Comment:
Question: What happens if we resume trials that already exist - are we currently saving the callbacks in the checkpoints (I don't think we are) and should we check if a trial is already registered with MLFlow?

Reply:
I guess the same question applies to Wandb...

Reply:
Yes that's a good point. This should work as long as the hash for the Trial remains the same. Like you said the same question applies to wandb- has there been any problems with wandb so far?